libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -40,48 +46,33 @@ public TranslationMiddleware(HttpClient httpClient, string[] nativeLanguages, st
                 if (!String.IsNullOrWhiteSpace(message.Text))
                 {
                     // determine the language we are using for this conversation
-                    var sourceLanguage = "en"; // context.Conversation.Data["Language"]?.ToString() ?? this.nativeLanguages.FirstOrDefault() ?? "en";
-
-                    var translationContext = new TranslationContext
+                    var sourceLanguage = await Task.Run(() => translator.Detect(message.Text)); // context.Conversation.Data["Language"]?.ToString() ?? this.nativeLanguages.FirstOrDefault() ?? "en";
+                    var templatePath = Path.Combine(new string[] { templatesDir, sourceLanguage + ".template" });
+                    if (File.Exists(templatePath))
+                    {
+                        this.translator.SetPostProcessorTemplate(templatePath);
+                    }
+                    if (!nativeLanguages.Contains(sourceLanguage))
                     {
-                        SourceText = message.Text,
-                        SourceLanguage = sourceLanguage,
-                        TargetLanguage = (this.nativeLanguages.Contains(sourceLanguage)) ? sourceLanguage : this.nativeLanguages.FirstOrDefault() ?? "en"
-                    };
-                    ((BotContext)context)["Translation"] = translationContext;
+                        var translationContext = new TranslationContext
+                        {
+                            SourceText = message.Text,
+                            SourceLanguage = sourceLanguage,
+                            TargetLanguage = (this.nativeLanguages.Contains(sourceLanguage)) ? sourceLanguage : this.nativeLanguages.FirstOrDefault() ?? "en"
+                        };
+                        ((BotContext)context)["Translation"] = translationContext;
```

Comment:
>((BotContext)context)["Translation"]  [](start = 24, length = 37)

we now have a Get() and Set() you can use versus the flex object.  You might also consider using something more GUID like ('microsoft.translation' or something) to avoid potential collisions.